openedx/core/djangoapps/content/learning_sequences/api/data.py

```diff
@@ -132,6 +138,10 @@ def not_deprecated(self, _attribute, value):
     # derived from what you pass into `sections`. Do not set this directly.
     sequences = attr.ib(type=Dict[UsageKey, CourseLearningSequenceData], init=False)
 
+    course_visibility = attr.ib(
+        type=str, validator=attr.validators.in_([visibility.value for visibility in list(CourseVisibilities)])
```

Comment:
Oh I see, you're explicitly extracting the string out of it. The thing about strings is that they're fuzzy and you can make mistakes that fail in quiet and not immediately obvious ways. Say someone later does a comparison somewhere like:

```python
if outline_data.course_visibility == 'public-outline':
   # stuff in this block will never get executed, but won't throw an error either.
   really_important_thing()
```

That string should be `'public_outline'` and not `'public-outline'`, but that code is not going to throw any kind of exception or do anything to let us know that it's mis-spelled. It will just quietly never enter that block. On the other hand, doing something like:

```python
if outline_data.course_visibility == CourseVisibilities.CORSE_OUTLINE:
   # kaboom!
   really_important_thing()
```

Doing the above comparison blows up because that enumeration attribute doesn't exist. Likewise, calling `CourseVisibilities('public_outline')` works but `CourseVisibilities('public-outline')` explodes with an exception.

So in general, please try to keep data in its stricter form in the `data.py` data model and convert to and from string at the edges–so for example, storing the enum directly, rather than the string value of it.

Reply:
Done. :+1: